### PR TITLE
docs(aio): add language service doc

### PR DESCRIPTION
**Pull request by kapunahelewong**
## PR Checklist
Does please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Adds a doc on how to use Angular Language Services.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Based on Chuck's ng-conf talk on Angular Language Services.

### CODE REVIEW

**Comment by kapunahelewong** [comment]
@chuckjaz Would you look this over to help me make sure I have my facts correct? I would especially like your thoughts on the latter half about how to create your own host and language service. Thank you!

File: aio/content/guide/language-service.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+yarn install
+```
+
+
+### Sublime Text
+
+In [Sublime Text](https://www.sublimetext.com/), you first need an extension to allow Typescript. Install the `@next` version of typescript in a local `node_modules` directory:
+
+```sh
+npm install --save-dev typescript@next
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: Now that 2.4 is released, you just need 2.4+ version of TypeScript so just installing latest TypeScript is sufficient.

File: aio/content/guide/language-service.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+```
+
+
+## Installing in your project
+
+You can also install Angular Language Service in your project with the 
+following `npm` command:
+
+```sh
+npm install --save-dev @angular/language-service
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: This also requires adding:

```js
    "plugins": [
        {"name": "@angular/language-service"}
    ]
```

to the `"compilerOptions"` section of the project's `tsconfig.json` file.

Note that this only provides diagnostics and completions in `.ts` files. A custom sublime plugin (or modifications to the current plugin) are required for completions in `.html` files.

File: aio/content/guide/language-service.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+Any time you type inside of the editor, it sends information to the other process to 
+track the state of your project. When you trigger a completion list within a template, the editor process first parses the template into an HTML AST, or [abstract syntax table](https://en.wikipedia.org/wiki/Abstract_syntax_tree). Then the Angular compiler interprets 
+what module the template is part of, the scope you're in, and the component selector. Then it figures out where in the template AST your cursor is. When it determines the 
+context, it can then determine what the children can be.
+
+It's a little more involved if you are in an interpolation. If you have an interpolation of `{{data.---}}` inside a `div` and need the completion list after `data.---`, the compiler can't use the HTML AST to find the answer. The HTML AST can only tell the compiler that there is some text with the characters "`{{data.---}}`". That's when the template parser produces an expression AST, which resides within the template AST. The Angular Language Service, which is within 
+the TypeScript Language Service, then looks at `data.---` within its context and determines
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: Consider, "The Angular Language Services ten looks at `data.---` within its context and asks the  TypeScript Language Service what the members of `data` are."

File: aio/content/guide/language-service.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+track the state of your project. When you trigger a completion list within a template, the editor process first parses the template into an HTML AST, or [abstract syntax table](https://en.wikipedia.org/wiki/Abstract_syntax_tree). Then the Angular compiler interprets 
+what module the template is part of, the scope you're in, and the component selector. Then it figures out where in the template AST your cursor is. When it determines the 
+context, it can then determine what the children can be.
+
+It's a little more involved if you are in an interpolation. If you have an interpolation of `{{data.---}}` inside a `div` and need the completion list after `data.---`, the compiler can't use the HTML AST to find the answer. The HTML AST can only tell the compiler that there is some text with the characters "`{{data.---}}`". That's when the template parser produces an expression AST, which resides within the template AST. The Angular Language Service, which is within 
+the TypeScript Language Service, then looks at `data.---` within its context and determines 
+what the members of `data` are. TypeScript then returns the list of possibilities.
+
+### Integrate with your editor
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: Consider removing this section and replacing it with a pointer to the `types.ts` in the language service sources which is the API documentation for the Language Service.

If we want to keep this section notes below.

File: aio/content/guide/language-service.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+You can still use the language service if you prefer an editor not listed on this page.
+
+You'll need to implement the following `LanguageServiceHost` interface.
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: The use would almost never need to implement this as `TypeScriptServiceHost` below implements it.

It is provided if to be able to use the language service with languages other than TypeScript.

We should ignore it here. Just say to create the language service through `createLanguageServiceFromTypescript()` which takes a TypeScript language service instance and returns an Angular language service instance.

File: aio/content/guide/language-service.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+service itself and its host. The host uses TypeScript to determine things 
+like what should go into a completion list. 
+If you create your own `TypeScriptServiceHost`, you also need to 
+create your Angular Language Service. Pass in the host that you're 
+using for the TypeScript service and the service 
+itself so you get back a language service host. 
+
+```typescript
+
+export class TypeScriptServiceHost implements LanguageServiceHost {
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: This class is implemented in `@angular/langauge-service`

File: aio/content/guide/language-service.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+...
+}
+const ngHost = new TypeScriptServiceHost(host, service);
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: Recommend `createLanguageServiceFromTypeScript()` instead.

File: aio/content/guide/language-service.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+const diagnostics = ngService.getDiagnostics(...);
+
+```
+
+
+If you do create an editor integration, the one thing that the Angular 
+Language Service does that Typscript doesn't is provide diagnostics 
+and completions inside of an HTML file. Once you have a `LanguageServiceHost`, 
+you can tell the host to return all of the HTML files that 
+are part of the project that are templates.
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: The list of which `HTML` files are templates is returned by `LanguageService.getTemplateReferenes()`.

File: aio/content/guide/language-service.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+```typescript
+
+export class TypeScriptServiceHost implements LanguageServiceHost {
+...
+}
+const ngHost = new TypeScriptServiceHost(host, service);
+const ngService = createLanguageService(ngHost);
+...
+const diagnostics = ngService.getDiagnostics(...);
```
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: Consider referencing the language service documentation about the other methods of `LanguageService`.

**Review by chuckjaz — CHANGES_REQUESTED** [review]

File: aio/content/guide/language-service.md (hunk 1 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Fixed.

File: aio/content/guide/language-service.md (hunk 2 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Done.

File: aio/content/guide/language-service.md (hunk 3 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Changed.

File: aio/content/guide/language-service.md (hunk 4 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Taking your advice here. Referring to the API docs makes more sense for the below.

File: aio/content/guide/language-service.md (hunk 2 above)
Comment (chuckjaz) [inline aio/content/guide/language-service.md]: This is also required for sublime to work.

**Review by chuckjaz — APPROVED** [review]
One nit.

**Comment by kapunahelewong** [comment]
Thank you @chuckjaz. I fixed the last comment about Sublime along with one edit I'd unintentionally left in about the TypeScript version.

File: aio/content/guide/language-service.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+for completions in HTML files.
+
+
+## How the Language Service works
+
+When you use an editor with a language service, there's an 
+editor process which starts a separate language process/service 
+to which it speaks through an [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call). 
+Any time you type inside of the editor, it sends information to the other process to 
+track the state of your project. When you trigger a completion list within a template, the editor process first parses the template into an HTML AST, or [abstract syntax table](https://en.wikipedia.org/wiki/Abstract_syntax_tree). Then the Angular compiler interprets
```
Comment (gkalpak) [inline aio/content/guide/language-service.md]: _abstract syntax table_ --> _abstract syntax tree_

File: aio/content/guide/language-service.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+  width="100%">
+</figure>
+
+
+## Angular Language Service in your editor
+
+Angular Language Service is currently available for [Visual Studio Code](https://code.visualstudio.com/) and 
+[WebStorm](https://www.jetbrains.com/webstorm). 
+You can also use Angular Language Service in your code editor of choice by 
+[configuring your language host manually](guide/language-service#integrate-with-your-editor).
```
Comment (gkalpak) [inline aio/content/guide/language-service.md]: This anchor does not exist.

File: aio/content/guide/language-service.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+elements. Any elements you have as a component selector will 
+show up in the completion list.
+
+## Error checking
+
+The Angular Language Service can also forewarn you of mistakes in your code. 
+In this example, Angular doesn't know what `orders` is or where it comes from. 
+
+<figure>
+  <img src="generated/images/guide/language-service/language-error.gif" alt="error checking" width="100%">
```
Comment (gkalpak) [inline aio/content/guide/language-service.md]: The `width` attribute only accepts a number (which is interpretead as the width in pixels). `100%` shouldn't have any effect. If you really needs this to be 100% wide (which I don't think it a good idea), we probably need a new style for that (or maybe there is already one). @sjtrimble might know more.

File: aio/content/guide/language-service.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
+# Angular Language Service
+
+The Angular Language Service is a way to get completions, errors, 
+hints, and navigation inside your Angular templates whether they 
+are external in an HTML file or embedded in annotations/decorators 
+in a string. The Angular Language Service autodetects that you are 
+opening an Angular file, reads your `ts.config` file, finds all the
```
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Ha! Good catch, thank you!

File: aio/content/guide/language-service.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+in a string. The Angular Language Service autodetects that you are 
+opening an Angular file, reads your `ts.config` file, finds all the 
+templates you have in your application, and then provides language 
+services for any templates that you open.
+
+
+## Autocompletion
+
+Autocompletion can speed up your development time by providing you with 
+contextual possiblities and hints as you type. This example shows
```
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Fixed.

File: aio/content/guide/language-service.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+opening an Angular file, reads your `ts.config` file, finds all the 
+templates you have in your application, and then provides language 
+services for any templates that you open.
+
+
+## Autocompletion
+
+Autocompletion can speed up your development time by providing you with 
+contextual possiblities and hints as you type. This example shows 
+autocomplete in an interpolations. As you type it out,
```
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Fixed.

File: aio/content/guide/language-service.md (hunk 12 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Ah, thank you. At first I'd had some trouble with how they looked but now with it just taken out completely it looks fine.

File: aio/content/guide/language-service.md (hunk 11 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Good catch. I'd just deleted it and forgot to delete this link. Fixed.

File: aio/content/guide/language-service.md (hunk 10 above)
Comment (kapunahelewong) [inline aio/content/guide/language-service.md]: Fixed! Thank you so much, @gkalpak!

**Review by juleskremer — APPROVED** [review]

**Pull request merged**